src/zelos/emulator/base.py

```diff
@@ -14,14 +14,285 @@
+        Returns all data in the region.
+
+        Returns:
+            Data from the region.
+        """
+        return self.emu.mem_read(self.address, self.size)
+
+
+class PageTable:
+    """
+    Maps host system pages to guest system pages. Enables directly
+    reading and writing emulated memory without using the emulator
+    memory API.
+    """
+
+    PAGE_MASK = 0xFFFFFFFFFFFFF000
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self._pages = dict()
+
+    def add(self, section) -> None:
+        """
+        Add or replace pages from the given section to the page table.
+
+        Args:
+            section: the memory region to add.
+
+        """
+        for addr in range(
+            section.address, section.address + section.size, 0x1000
+        ):
+            self._pages[addr] = section
+
+    def remove(self, section) -> None:
+        """
+        Remove pages from the given section from the page table
+
+        Args:
+            section: the memory region to remove.
+
+        """
+        for addr in range(
+            section.address, section.address + section.size, 0x1000
+        ):
+            del self._pages[addr]
+
+    def exists(self, address: int) -> bool:
+        """
+        Checks if the address exists in the page table.
+
+        Args:
+            address: The address of the page to check.
+
+        Returns:
+            True if the address exists in the page table, False
+                otherwise.
+
+        """
+        return address & PageTable.PAGE_MASK in self._pages
+
+    def read(self, address: int, size: int) -> bytearray:
+        """
+        Reads `size` bytes of guest memory from `address`.
+
+        Args:
+            address: The address to read.
+            size: The size of the data to fetch in bytes.
+
+        Returns:
+            The data at the specified address.
+
+        """
+        if size == 0:
+            return bytearray()
+        page_addr = address & PageTable.PAGE_MASK
+
+        # Fast path if all data is on one page.
+        try:
+            mr = self._pages[page_addr]
+            offset = address - mr.address
+            if offset + size < mr.size:
+                return mr.host_data[offset : offset + size]
+        except (IndexError, ValueError):
+            pass
+        except KeyError:
+            raise MemoryReadUnmapped(
+                f"Read unmapped memory at address 0x{address:x}"
+            )
+
+        # Slow path if data spans multiple pages.
+        data = bytearray()
+        while size > 0:
+            try:
+                mr = self._pages[address & PageTable.PAGE_MASK]
+            except KeyError:
+                raise MemoryReadUnmapped(
+                    f"Read unmapped memory at address 0x{address:x}"
+                )
+            offset = address - mr.address
+            bytes_to_read = min(size, mr.size - offset)
+            data += mr.host_data[offset : offset + size]
```

Comment:
This would be easier to read if 
`data += mr.host_data[offset : offset + size]` -> `data += mr.host_data[offset : offset + bytes_to_read]`
Those should be equivalent?